cask/artifact/symlinked: do not overwrite files or symbolic links

miccal opened the pull request:
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

When installing a Cask that includes a `binary` or a `manpage` there is the possibility that there already exists either a file or a symbolic link in `/usr/local/bin` or `/usr/local/share/man/*`.

At present, `brew install {some cask with a binary or a manpage}` will _not_ overwrite a file in `/usr/local/bin` or `/usr/local/share/man/*`, but _will_ overwrite a symbolic link in `/usr/local/bin` or `/usr/local/share/man/*`.

This change ensures that neither a file nor a symbolic link will be overwritten.

See [the issue here in `homebrew-cask`](https://github.com/Homebrew/homebrew-cask/issues/87304) for more background.

Thank you.

vitorgalvao commented on Library/Homebrew/cask/artifact/symlinked.rb:
```diff
@@ -44,10 +44,11 @@ def link(**options)
-                "#{self.class.english_name} at '#{target}'; not linking."
+                "It seems there already exists #{self.class.english_article} " \
+                "#{self.class.english_name} at '#{target}'; not overwriting. " \
+                "To install the Cask use the #{Formatter.identifier("--no-binaries")} option."
```
Still sounds like a recommendation, as if it were a circumventable bug. It should be clear it’s a tradeoff users should make a decision on. Something like “To skip the installation of the CLI tools, use the `--no-binaries` flag”.

miccal replied:
Thanks, I have made your suggested change but left the wording as `option` instead of `flag` as that is what they are mostly referred to in `man brew`.

miccal commented:
This change causes one of the `brew tests` to fail, namely:
```
rspec ./test/cask/artifact/binary_spec.rb:70 # Cask::Artifact::Binary clobbers an existing symlink
```

reitermarkus commented:
This should be in `Artifact::Binary` since `--no-binaries` only applies to that.

miccal commented:
> This should be in `Artifact::Binary` since `--no-binaries` only applies to that.

Sorry @reitermarkus, but could you explain a bit more what you mean?

reitermarkus commented:
The message should only be shown for `binary` artifacts, since those are the only ones affected by the `--no-binaries` flag. So the message should be in the `Binary` class instead of the `Symlinked` class.

miccal commented:
Please be patient while I think and work on this ... thanks.

miccal commented:
> This should be in Artifact::Binary since --no-binaries only applies to that.

Since the `Symlinked` class is used by both the `Binary` class and the `Manpage` class, I have changed the message to be more generic and make no mention of the `--no-binaries` flag.

I have updated my explanation for this change in the fist post above.

I have tested this change locally by installing a fake cask with a `binary` and a `manpage` and can confirm the error:
```
Error: It seems there already exists a {Binary,Manpage} at '{/usr/local/bin/CONFLICT,/usr/local/share/man/CONFLICT}'; not overwriting.
```
regardless of whether `CONFLICT` is a file or a symbolic link.

However, I am not sure how to fix the following `brew tests` error:
```
|-> brew tests --only=cask/artifact/binary --verbose --debug
Randomized with seed 11751
bundle exec parallel_rspec --nice -- -I /usr/local/Homebrew/Library/Homebrew/test --seed 11751 --color --require spec_helper --format NoSeedProgressFormatter --format ParallelTests::RSpec::RuntimeLogger --out /Users/miccal/Library/Caches/Homebrew/tests/parallel_runtime_rspec.log --tag ~needs_linux -- test/cask/artifact/binary_spec.rb
1 processes for 1 specs, ~ 1 specs per process
F......

Failures:

  1) Cask::Artifact::Binary clobbers an existing symlink
     Failure/Error: artifact.install_phase(command: NeverSudoSystemCommand, force: false)
     
     Cask::CaskError:
       It seems there already exists a Binary at '/tmp/homebrew-tests-20200817-32586-4ozcmq/prefix/bin/binary'; not overwriting.
     # ./cask/artifact/symlinked.rb:48:in `link'
     # ./cask/artifact/binary.rb:9:in `link'
     # ./cask/artifact/symlinked.rb:17:in `install_phase'
     # ./test/cask/artifact/binary_spec.rb:74:in `block (3 levels) in <top (required)>'
     # ./test/cask/artifact/binary_spec.rb:73:in `each'
     # ./test/cask/artifact/binary_spec.rb:73:in `block (2 levels) in <top (required)>'
     # ./test/support/helper/spec/shared_context/homebrew_cask.rb:51:in `block (2 levels) in <top (required)>'
     # ./test/spec_helper.rb:189:in `block (3 levels) in <top (required)>'
     # ./test/spec_helper.rb:188:in `block (2 levels) in <top (required)>'
     # ./vendor/bundle/ruby/2.6.0/gems/rspec-retry-0.6.2/lib/rspec/retry.rb:124:in `block in run'
     # ./vendor/bundle/ruby/2.6.0/gems/rspec-retry-0.6.2/lib/rspec/retry.rb:110:in `loop'
     # ./vendor/bundle/ruby/2.6.0/gems/rspec-retry-0.6.2/lib/rspec/retry.rb:110:in `run'
     # ./vendor/bundle/ruby/2.6.0/gems/rspec-retry-0.6.2/lib/rspec_ext/rspec_ext.rb:12:in `run_with_retry'
     # ./vendor/bundle/ruby/2.6.0/gems/rspec-retry-0.6.2/lib/rspec/retry.rb:37:in `block (2 levels) in setup'
     # ./vendor/bundle/ruby/2.6.0/gems/rspec-wait-0.0.9/lib/rspec/wait.rb:46:in `block (2 levels) in <top (required)>'

Finished in 7.44 seconds (files took 1.55 seconds to load)
7 examples, 1 failure

Failed examples:

rspec ./test/cask/artifact/binary_spec.rb:70 # Cask::Artifact::Binary clobbers an existing symlink


7 examples, 1 failure

Took 9 seconds
Tests Failed
```

miccal commented:
I think I figured it out ... ping @reitermarkus and @vitorgalvao.

reitermarkus commented on Library/Homebrew/test/cask/artifact/binary_spec.rb:
```diff
@@ -70,11 +70,13 @@
   it "clobbers an existing symlink" do
```
“does not …”

SeekingMeaning replied:
or “avoids clobbering …” to match the test above this one

miccal replied:
Fixed, thank you @reitermarkus and @SeekingMeaning.

Stupid oversight on my part.

miccal commented:
Thank you @reitermarkus.
